File: Board/src/main/java/com/example/board/domain/user/facade/UserFacade.java
package com.example.board.domain.user.facade;

import com.example.board.domain.user.domain.User;
import com.example.board.domain.user.domain.repository.UserRepository;
import com.example.board.domain.auth.exception.PasswordMismatchedException;
import com.example.board.domain.user.exception.UserNotFoundException;
import lombok.RequiredArgsConstructor;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Component;

@RequiredArgsConstructor
@Component
public class UserFacade { //Facade Pattern(Design Pattern) : 서브시스템에 있는 인터페이스들에 대한 통합된 인터페이스를 제공 -> 코드 재사용성 증가
    private final UserRepository userRepository;
    private final PasswordEncoder passwordEncoder;

    public User getCurrentUser() {
        String accountId = SecurityContextHolder.getContext().getAuthentication().getName(); //SecurityContextHolder : SecurityContext 저장소(꼭 구글링해보기)

        return userRepository.findByAccountId(accountId)
                .orElseThrow(() -> UserNotFoundException.EXCEPTION);
    }

    public void checkPassword(User user, String password) {
        if(!passwordEncoder.matches(password, user.getPassword())) {
            throw PasswordMismatchedException.EXCEPTION;
        }
    }

    public Boolean checkUserExist(String accountId) {
        return userRepository.existsByAccountId(accountId);
    }

    public Long getUserId() {
        return getCurrentUser().getId();
    }
}

File: Kopring/Dise_V2/src/main/kotlin/com/example/dise_v2/domain/user/facade/UserFacade.kt
package com.example.dise_v2.domain.user.facade

import com.example.dise_v2.domain.user.domain.User
import com.example.dise_v2.domain.user.domain.repository.UserRepository
import com.example.dise_v2.domain.user.exception.AlreadyUserExistException
import com.example.dise_v2.domain.user.exception.PasswordMisMatshException
import com.example.dise_v2.domain.user.exception.UserNotFoundException
import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.stereotype.Component

@Component
class UserFacade(
    private val userRepository: UserRepository,
    private val passwordEncoder: PasswordEncoder
) {
    fun getCurrentUser(): User {
        val accountId: String = SecurityContextHolder.getContext().authentication.name

        return getByAccountId(accountId)
    }

    fun checkUserExist(accountId: String): Boolean {
        return userRepository.existsByAccountId(accountId)
    }

    fun checkPassword(user: User, passsword: String) {
        if (!passwordEncoder.matches(passsword, user.password)) {
            throw PasswordMisMatshException.EXCEPTION
        }
    }

    fun getByAccountId(accountId: String): User {
        return userRepository.findByAccountId(accountId) ?: throw UserNotFoundException.EXCEPTION
    }
}